zipline/pipeline/engine.py

```diff
@@ -881,3 +877,34 @@ def _resolve_hooks(self, hooks):
         if hooks is None:
             hooks = []
         return DelegatingHooks(self._default_hooks + hooks)
+
+
+def _pipeline_output_index(dates, assets, mask):
+    """
+    Create a MultiIndex for a pipeline output.
+
+    Parameters
+    ----------
+    dates : pd.DatetimeIndex
+        Row labels for ``mask``.
+    assets : pd.Index
+        Column labels for ``mask``.
+    mask : np.ndarray[bool]
+        Mask array indicating date/asset pairs that should be included in
+        output index.
+
+    Returns
+    -------
+    index : pd.MultiIndex
+        MultiIndex  containing (date,  asset) pairs  corresponding to  ``True``
+        values in ``mask``.
+    """
+    date_labels = repeat_last_axis(arange(len(dates)), len(assets))[mask]
+    asset_labels = repeat_first_axis(arange(len(assets)), len(dates))[mask]
+    return MultiIndex(
+        levels=[dates, assets],
+        labels=[date_labels, asset_labels],
+        # TODO: We should probably add names for these.
```

Comment:
We might want to consider putting `assets` and `dates` as the column labels here, but I didn't want to do that in this change to minimize functional changes. (We also have some test suites that break with that change.)